src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs

```diff
@@ -265,46 +293,130 @@ private BoundExpression MakePatternIndexOffsetExpression(BoundExpression unlower
             MethodSymbol sliceMethod,
             BoundExpression rangeArg)
         {
-            // Lowered code:
+            Debug.Assert(TypeSymbol.Equals(
+                rangeArg.Type,
+                _compilation.GetWellKnownType(WellKnownType.System_Range),
+                TypeCompareKind.ConsiderEverything));
+
+            // Lowered code without optimizations:
             // var receiver = receiverExpr;
             // int length = receiver.length;
             // Range range = argumentExpr;
             // int start = range.Start.GetOffset(length)
-            // int end = range.End.GetOffset(length)
-            // receiver.Slice(start, end - start)
+            // int rangeSize = range.End.GetOffset(length) - start
+            // receiver.Slice(start, rangeSize)
 
             var F = _factory;
 
+            var localsBuilder = ArrayBuilder<LocalSymbol>.GetInstance();
+            var sideEffectsBuilder = ArrayBuilder<BoundExpression>.GetInstance();
+
             var receiverLocal = F.StoreToTemp(VisitExpression(receiver), out var receiverStore);
             var lengthLocal = F.StoreToTemp(F.Property(receiverLocal, lengthOrCountProperty), out var lengthStore);
-            var rangeLocal = F.StoreToTemp(VisitExpression(rangeArg), out var rangeStore);
-            var startLocal = F.StoreToTemp(
-                F.Call(
-                    F.Call(rangeLocal, F.WellKnownMethod(WellKnownMember.System_Range__get_Start)),
-                    F.WellKnownMethod(WellKnownMember.System_Index__GetOffset),
-                    lengthLocal),
-                out var startStore);
-            var endLocal = F.StoreToTemp(
-                F.Call(
-                    F.Call(rangeLocal, F.WellKnownMethod(WellKnownMember.System_Range__get_End)),
-                    F.WellKnownMethod(WellKnownMember.System_Index__GetOffset),
-                    lengthLocal),
-                out var endStore);
+
+            localsBuilder.Add(receiverLocal.LocalSymbol);
+            sideEffectsBuilder.Add(receiverStore);
+
+            BoundExpression startExpr;
+            BoundExpression rangeSizeExpr;
+            if (rangeArg is BoundRangeExpression rangeExpr)
```

Comment:
>if [](start = 12, length = 2)

The test cases I would expect for the new/optimized code: pattern indexer with `..`, `..y`, `x..` and `x..y`. I wasn't able to identify test(s) for all of them. Can you confirm the coverage?

Reply:
I looked through and found a test for everything except `..`. I've added one in the latest commit.